source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,194 @@
+// CacheFilter::encodeData.
+class CacheInsertFragmentBody : public CacheInsertFragment {
+public:
+  CacheInsertFragmentBody(const Buffer::Instance& buffer, bool end_stream)
+      : buffer_(buffer), end_stream_(end_stream) {}
+
+  void
+  send(InsertContext& context,
+       std::function<void(bool cache_success, bool end_stream, size_t sz)> on_complete) override {
+    size_t sz = buffer_.length();
+    context.insertBody(
+        std::move(buffer_),
+        [on_complete, end_stream = end_stream_, sz](bool cache_success) {
+          on_complete(cache_success, end_stream, sz);
+        },
+        end_stream_);
+  }
+
+private:
+  Buffer::OwnedImpl buffer_;
+  const bool end_stream_;
+};
+
+// A CacheInsertFragment containing the full trailers of the response.
+class CacheInsertFragmentTrailers : public CacheInsertFragment {
+public:
+  explicit CacheInsertFragmentTrailers(const Http::ResponseTrailerMap& trailers)
+      : trailers_(Http::ResponseTrailerMapImpl::create()) {
+    Http::ResponseTrailerMapImpl::copyFrom(*trailers_, trailers);
+  }
+
+  void
+  send(InsertContext& context,
+       std::function<void(bool cache_success, bool end_stream, size_t sz)> on_complete) override {
+    // While zero isn't technically true for the size of trailers, it doesn't
+    // matter at this point because watermarks after the stream is complete
+    // aren't useful.
+    context.insertTrailers(
+        *trailers_, [on_complete](bool cache_success) { on_complete(cache_success, true, 0); });
+  }
+
+private:
+  std::unique_ptr<Http::ResponseTrailerMap> trailers_;
+};
+
+CacheInsertQueue::CacheInsertQueue(Http::StreamEncoderFilterCallbacks& encoder_callbacks,
+                                   InsertContextPtr insert_context, AbortInsertCallback abort)
+    : dispatcher_(encoder_callbacks.dispatcher()), insert_context_(std::move(insert_context)),
+      low_watermark_bytes_(encoder_callbacks.encoderBufferLimit() / 2),
+      high_watermark_bytes_(encoder_callbacks.encoderBufferLimit()),
+      encoder_callbacks_(encoder_callbacks), abort_callback_(abort) {}
+
+void CacheInsertQueue::insertHeaders(const Http::ResponseHeaderMap& response_headers,
+                                     const ResponseMetadata& metadata, bool end_stream) {
+  end_stream_queued_ = end_stream;
+  // While zero isn't technically true for the size of headers, headers are
+  // typically excluded from the stream buffer limit.
+  insert_context_->insertHeaders(
+      response_headers, metadata,
+      [this, end_stream](bool cache_success) { onFragmentComplete(cache_success, end_stream, 0); },
+      end_stream);
+  fragment_in_flight_ = true;
+}
+
+void CacheInsertQueue::insertBody(const Buffer::Instance& fragment, bool end_stream) {
+  if (end_stream) {
+    end_stream_queued_ = true;
+  }
+  if (fragment_in_flight_) {
+    size_t sz = fragment.length();
+    queue_size_bytes_ += sz;
+    fragments_.push_back(std::make_unique<CacheInsertFragmentBody>(fragment, end_stream));
```

Comment:
WDYT about fragmentation in buffers here? I'm not too familiar with this exact filter -- but my concern is around multiple "Buffers" entering this list that hold a trivial amount of data but are fragmented.

Reply:
I would not like addressing that to be part of this change. The buffers that are passed to the cache implementation *before* this change are the same buffers as *after* this change, just with a difference of timing. I would rather changing the expectations around that be a separate operation, if at all.

Reply:
Do the `insertBody()` calls prior keep seperate buffers or are they merged? If not merged, then this timing could be impactful?

Reply:
The impact of the timing change (on the file system cache implementation at least!) is covered by the existing tests, and also right now largely irrelevant because it's just going through this queue straight into another identical queue that belongs to the cache implementation. The *next* change after this is cleaning that up so the cache implementation doesn't own a queue any more. That other queue was a horrible workaround for the cache filter previously not actually implementing its own described API. Any existing cache implementation that *doesn't* currently have a queue must be not-asynchronous (because async caching without a queue was broken) and therefore should not be meaningfully impacted by this change.

However, I think if we changed this to merge buffers in the queue it would have a good chance of breaking existing tests that expect the encodeData inputs to match what the cache instance sees. I agree it might be a nice optimization to merge the buffers, though for the most part I would expect it to make no practical difference (my expectation is that *usually*, in practice, the queue would drain as fast as it populates). I'll make a ticket for that too.

Reply:
> also right now largely irrelevant because it's just going through this queue straight into another identical queue that belongs to the cache implementation
The implementation of this queue here could be more inefficient than cache queue implementation e.g. the cache queue that is a buffer for example vs this deque that has fragmentation issues.

sgtm making a ticket for this. While _usually_ this queue will drain as fast as possible when everything is working, there will definitely exist times when they aren't and this suffers from fragmentation.

Reply:
#28813